polkadot-test-service/src/lib.rs

```diff
@@ -0,0 +1,316 @@
+		max_runtime_instances: 8,
+		announce_block: true,
+		base_path: Some(base_path),
+		informant_output_format,
+	})
+}
+
+/// Run a Polkadot test node using the Polkadot test runtime. The node will be using an in-memory socket, therefore you
+/// need to provide boot nodes if you want it to be connected to other nodes. The `storage_update_func` can be used to
+/// make adjustements to the runtime before the node starts.
+pub fn run_test_node(
+	task_executor: TaskExecutor,
+	key: Sr25519Keyring,
+	storage_update_func: impl Fn(),
+	boot_nodes: Vec<MultiaddrWithPeerId>,
+) -> Result<
+	PolkadotTestNode<
+		TaskManager,
+		impl PolkadotClient<Block, TFullBackend<Block>, polkadot_test_runtime::RuntimeApi>,
+	>,
+	ServiceError,
+> {
+	let config = node_config(storage_update_func, task_executor, key, boot_nodes)?;
+	let multiaddr = config.network.listen_addresses[0].clone();
+	let authority_discovery_enabled = false;
+	let (task_manager, client, handles, network, rpc_handlers) =
+		polkadot_test_new_full(config, None, None, authority_discovery_enabled, 6000)?;
+
+	let peer_id = network.local_peer_id().clone();
+	let multiaddr_with_peer_id = MultiaddrWithPeerId { multiaddr, peer_id };
+
+	let node = PolkadotTestNode {
+		task_manager,
+		client,
+		handles,
+		multiaddr_with_peer_id,
+		rpc_handlers,
+	};
+
+	Ok(node)
+}
+
+/// A Polkadot test node instance used for testing.
+pub struct PolkadotTestNode<S, C> {
+	/// TaskManager's instance.
+	pub task_manager: S,
+	/// Client's instance.
+	pub client: Arc<C>,
+	/// Node's handles.
+	pub handles: FullNodeHandles,
+	/// The `MultiaddrWithPeerId` to this node. This is useful if you want to pass it as "boot node" to other nodes.
+	pub multiaddr_with_peer_id: MultiaddrWithPeerId,
+	/// RPCHandlers to make RPC queries.
+	pub rpc_handlers: Arc<RpcHandlers>,
+}
+
+impl<S, C> PolkadotTestNode<S, C>
+where
+	C: BlockchainEvents<Block>,
+{
+	/// Wait for `count` blocks to be imported in the node and then exit. This function will not return if no blocks
+	/// are ever created, thus you should restrict the maximum amount of time of the test execution.
+	pub fn wait_for_blocks(&self, count: usize) -> impl Future<Output = ()> {
+		assert_ne!(count, 0, "'count' argument must be greater than 1");
+		let client = self.client.clone();
+
+		wait_for_blocks(client.clone(), count)
+	}
+
+	/// Send a transaction through the RPCHandlers.
+	pub fn send_transaction(
+		&self,
+		extrinsic: OpaqueExtrinsic,
+	) -> impl Future<
+		Output = (
+			Option<String>,
+			RpcSession,
+			futures01::sync::mpsc::Receiver<String>,
+		),
+	> {
+		let rpc_handlers = self.rpc_handlers.clone();
+
+		send_transaction(rpc_handlers, extrinsic)
+	}
+}
+
+/// Wait for `count` blocks to be imported in the node and then exit. This function will not return if no blocks
+/// are ever created, thus you should restrict the maximum amount of time of the test execution.
+pub async fn wait_for_blocks<C, B>(client: Arc<C>, count: usize)
```

Comment:
And this one?

Reply:
Ahh they are, but why are they also free standing functions?

Can the struct function not also be `async`?

Reply:
I will talk more about that in another comment. I think there is something nicer to do!

> Can the struct function not also be async?

I clone the input before moving it to a future as they don't consume self.